controlplane/kubeadm/internal/etcd_client_generator.go

```diff
@@ -72,26 +78,48 @@ func (c *EtcdClientGenerator) forFirstAvailableNode(ctx context.Context, nodeNam
 
 // forLeader takes a list of nodes and returns a client to the leader node.
 func (c *EtcdClientGenerator) forLeader(ctx context.Context, nodeNames []string) (*etcd.Client, error) {
-	var errs []error
+	// This is an additional safeguard for avoiding this func to return nil, nil.
+	if len(nodeNames) == 0 {
+		return nil, errors.New("invalid argument: forLeader can't be called with an empty list of nodes")
+	}
 
+	// Loop through the existing control plane nodes.
+	var errs = make([]error, 0)
 	for _, nodeName := range nodeNames {
+		// Get a temporary client to the etcd instance hosted on the node.
 		client, err := c.forFirstAvailableNode(ctx, []string{nodeName})
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
 		defer client.Close()
+
+		// Get the list of members.
 		members, err := client.Members(ctx)
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
+
+		// Loop through the list of etcd members reported checking for the current leader.
 		for _, member := range members {
-			if member.Name == nodeName && member.ID == client.LeaderID {
-				return c.forFirstAvailableNode(ctx, []string{nodeName})
+			if member.ID == client.LeaderID {
+				// Gets a connection to the etcd leader via the node hosting it.
+				for _, nodeName := range nodeNames {
+					if member.Name == nodeName {
+						return c.forFirstAvailableNode(ctx, []string{nodeName})
+					}
+				}
 			}
 		}
-	}
 
+		// If it is not possible to get a connection to the leader via existing nodes,
+		// it means that the control plane is an invalid state, with an etcd member - the current leader -
+		// without a corresponding node.
+		// TODO: In future we can eventually try to automatically remediate this condition by moving the leader
+		//  to another member with a corresponding node.
+		return nil, errors.Errorf("could not find the node corresponding to the current etcd leader %x", client.LeaderID)
+		break
+	}
 	return nil, errors.Wrap(kerrors.NewAggregate(errs), "could not establish a connection to the etcd leader")
```

Comment:
Before this line we should always append to `errs` that we were not able to generate a client

Reply:
I can do this but it is redundant given that
- we are sure that there is at least one node
- when looping through nodes, either they return with a client or they generate an error

Reply:
I was suggesting it because it ensures that we'd never return a `nil` error even if the code above changes